libraries/standard/mqtt/src/mqtt.c

```diff
@@ -439,56 +445,87 @@ static bool matchWildcards( const char * pTopicName,
     assert( pTopicFilter != NULL );
     assert( topicFilterLength != 0 );
     assert( pNameIndex != NULL );
+    assert( pFilterIndex != NULL );
     assert( pMatch != NULL );
 
     /* Wild card in a topic filter is only valid either at the starting position
      * or when it is preceded by a '/'.*/
-    locationIsValidForWildcard = ( ( filterIndex == 0u ) ||
-                                   ( pTopicFilter[ filterIndex - 1U ] == '/' )
+    locationIsValidForWildcard = ( ( *pFilterIndex == 0u ) ||
+                                   ( pTopicFilter[ *pFilterIndex - 1U ] == '/' )
                                    ) ? true : false;
 
-    if( locationIsValidForWildcard == true )
+    if( ( pTopicFilter[ *pFilterIndex ] == '+' ) && ( locationIsValidForWildcard == true ) )
     {
-        if( pTopicFilter[ filterIndex ] == '+' )
+        bool nextLevelExistsInTopicName = false;
+        bool nextLevelExistsinTopicFilter = false;
+
+        /* Move topic name index to the end of the current level. The end of the
+         * current level is identified by the last character before the next level
+         * separator '/'. */
+        while( *pNameIndex < topicNameLength )
         {
-            /* Move topic name index to the end of the current level. The end of the
-             * current level is identified by '/'. */
-            while( ( *pNameIndex < topicNameLength ) && ( pTopicName[ *pNameIndex ] != '/' ) )
+            /* Exit the loop if we hit the level separator. */
+            if( pTopicName[ *pNameIndex ] == '/' )
             {
-                ( *pNameIndex )++;
+                nextLevelExistsInTopicName = true;
+                break;
             }
 
-            /* Decrement the topic name index for 2 different cases:
-             * - If the break condition is ( *pNameIndex < topicNameLength ), then
-             *   we have reached past the end of the topic name and we move back the
-             *   the index on the last character.
-             * - If the break condition is ( pTopicName[ *pNameIndex ] != '/' ), we
-             *   move back the index on the '/' character. */
-            ( *pNameIndex )--;
+            ( *pNameIndex )++;
         }
 
-        /* '#' matches everything remaining in the topic name. It must be the
-         * last character in a topic filter. */
-        else if( ( pTopicFilter[ filterIndex ] == '#' ) &&
-                 ( filterIndex == ( topicFilterLength - 1U ) ) )
+        /* Determine if the topic filter contains a child level after the current level
+         * represented by the '+' wildcard. */
+        if( ( *pFilterIndex < ( topicFilterLength - 1U ) ) &&
```

Comment:
If **pTopicFilter** = ['+'] and **topicFilterLength** = 1 and **\*pFilterIndex** = 0
Then pTopicFilter[ *pFilterIndex + 1U ] is an out of bounds reference if pTopicFilter is not NULL terminated.

Reply:
Oh nevermind sorry I see that this is < and not <= ignore my comment.